File: src/main/scala/utn/frba/iasc/injection/ServiceModule.scala
package utn.frba.iasc.injection

import akka.actor.{ActorRef, ActorSystem}
import com.google.inject.name.Named
import com.google.inject.{AbstractModule, Provides, Singleton}
import utn.frba.iasc.service.{AuctionService, BidService, UserService}

import scala.concurrent.ExecutionContextExecutor

object ServiceModule extends AbstractModule {
  @Provides
  @Singleton
  @Named("auctionService")
  def auctionService(
    @Named("routerActorRef") routerActor: ActorRef,
    @Named("executionContext") executionContext: ExecutionContextExecutor
  ): AuctionService = new AuctionService(
    babylonActor = routerActor,
    executionContext = executionContext
  )

  @Provides
  @Singleton
  @Named("userService")
  def userService(
    @Named("usersActorRef") usersActor: ActorRef,
    @Named("actorSystem") system: ActorSystem
  ): UserService = new UserService(
    babylonActor = usersActor,
    executionContext = system.dispatcher
  )

  @Provides
  @Singleton
  @Named("bidService")
  def bidService(
    @Named("routerActorRef") routerActor: ActorRef,
    @Named("executionContext") executionContextExecutor: ExecutionContextExecutor
  ): BidService = new BidService(
    routerActor = routerActor,
    executionContext = executionContextExecutor
  )
}

File: babylon-kt/babylon-out/src/main/kotlin/utn/frba/iasc/babylon/config/ServiceModule.kt
package utn.frba.iasc.babylon.config

import akka.actor.ActorRef
import com.google.inject.AbstractModule
import com.google.inject.Provides
import com.google.inject.Singleton
import com.google.inject.name.Named
import utn.frba.iasc.babylon.service.CallbackService

object ServiceModule : AbstractModule() {
    @Provides
    @Singleton
    @Named("callbackService")
    fun callbackService(
        @Named("callbackActor") callbackActor: ActorRef
    ) = CallbackService(callbackActor)
}